### PR TITLE
Detect capturing in ios

### PR DESCRIPTION
**Pull request by qliqdev**
Created new branch for one feature `detectCapturing`

### CODE REVIEW

**Comment by robingenz** [comment]
I suggest we start with a simple Capacitor Plugin [Event Listener](https://capacitorjs.com/docs/plugins/ios#plugin-events) for detecting screenshots and leave the functionality to the app:

```ts
/**
   * Listen for screen captures.
   * 
   * Only available on iOS.
   */
  addListener(
    eventName: 'capturingDetected',
    listenerFunc: () => void,
  ): Promise<PluginListenerHandle> & PluginListenerHandle;
  /**
   * Remove all listeners for this plugin.
   */
  removeAllListeners(): Promise<void>;
```

I can imagine that not every user of the plugin wants this functionality. Maybe screenshots should be allowed despite the current privacy screen.

**Comment by qliqdev** [comment]
Sure, let me fix it today.

**Comment by qliqdev** [comment]
> Sure, let me fix it today.

Build fails. is `project.pbxproj ` ok?

File: ios/Plugin/PrivacyScreenPlugin.swift [hunk 1]
```diff
@@ -38,7 +38,7 @@ public class PrivacyScreenPlugin: CAPPlugin {
         call.resolve()
     }
 
-    @objc func present(_ call: CAPPluginCall) {
+    @objc func show(_ call: CAPPluginCall) {
```
Comment (qliqdev) [inline ios/Plugin/PrivacyScreenPlugin.swift:41]: was renamed, cuz  `present` is already exist

**Review by robingenz — CHANGES_REQUESTED** [review]
Thank you for the changes!

File: src/definitions.ts [hunk 2]
```diff
@@ -36,4 +48,41 @@ export interface PrivacyScreenPlugin {
...
+  show(): Promise<void>;
+
+  /**
+   * Dismiss the privacy screen protection.
+   *
+   * Only available for iOS.
+   *
+   * @since 3.0.2
+   */
+  hide(): Promise<void>;
```
[the remark this answers is not in the document]
Comment (qliqdev) [reply src/definitions.ts:68]: Sure, i think this part can be removed.

File: README.md [hunk 3]
```diff
@@ -101,6 +101,16 @@ const enable = async () => {
...
+
+const detect = async () => {
+  await PrivacyScreen.addListener('capturingDetected', async (capturing) => {
+    if (capturing.capturing) {
+      await PrivacyScreen.show();
+    } else {
+      await PrivacyScreen.hide();
+    }
+  });
+}
```
Comment (robingenz) [inline README.md:113]: Please remove this outdated example.

File: README.md [hunk 4]
```diff
@@ -109,6 +119,11 @@ const disable = async () => {
 
 * [`enable()`](#enable)
 * [`disable()`](#disable)
+* [`show()`](#show)
+* [`hide()`](#hide)
+* [`addListener('capturingDetected', ...)`](#addlistenercapturingdetected)
```
Comment (robingenz) [inline README.md:124]: Please run `npm run docgen` to update the `README.md`.

File: ios/Plugin/PrivacyScreenPlugin.swift [hunk 5]
```diff
@@ -36,20 +42,36 @@ public class PrivacyScreenPlugin: CAPPlugin {
...
         DispatchQueue.main.async {
             self.bridge?.viewController?.present(self.privacyViewController!, animated: false, completion: nil)
         }
     }
 
     @objc func onAppDidBecomeActive() {
+        guard self.isEnabled else {
+            return
+        }
+
```
Comment (robingenz) [inline ios/Plugin/PrivacyScreenPlugin.swift:55]: The privacy view should always be dismissed when the app becomes active. Please revert this changes.

File: ios/Plugin/PrivacyScreenPlugin.swift [hunk 6]
```diff
@@ -36,20 +42,36 @@ public class PrivacyScreenPlugin: CAPPlugin {
...
     }
 
+    @objc private func onAppDetectCapturing() {
+        guard self.isEnabled else {
+            return
+        }
+
+        if #available(iOS 11.0, *) {
+            if(UIScreen.main.isCaptured){
+                self.notifyListeners("screenRecordingStarted", data: [:])
```
Comment (robingenz) [inline ios/Plugin/PrivacyScreenPlugin.swift:68]:
```suggestion
                self.notifyListeners("screenRecordingStarted", data: nil)
```

File: ios/Plugin/PrivacyScreenPlugin.swift [hunk 7]
```diff
@@ -36,20 +42,36 @@ public class PrivacyScreenPlugin: CAPPlugin {
...
+            } else {
+                self.notifyListeners("screenRecordingStopped", data: [:])
```
Comment (robingenz) [inline ios/Plugin/PrivacyScreenPlugin.swift:70]:
```suggestion
                self.notifyListeners("screenRecordingStopped", data: nil)
```

File: src/definitions.ts [hunk 8]
```diff
@@ -36,4 +39,35 @@ export interface PrivacyScreenPlugin {
    * @since 1.1.0
    */
   disable(): Promise<void>;
+
+  /**
+   * Listen for screen started recording.
```
Comment (robingenz) [inline src/definitions.ts:44]:
```suggestion
   * Called when the screen recording is started.
```

File: src/definitions.ts [hunk 9]
```diff
@@ -36,4 +39,35 @@ export interface PrivacyScreenPlugin {
...
+   *
+   * @since 3.0.2
+   */
+  addListener(
+    eventName: 'screenRecordingStarted',
+    listenerFunc: () => void,
+  ): Promise<PluginListenerHandle> & PluginListenerHandle;
+
+  /**
+   * Listen for screen stopped recording.
```
Comment (robingenz) [inline src/definitions.ts:56]:
```suggestion
   * Called when the screen recording is stopped.
```

**Review by robingenz — CHANGES_REQUESTED** [review]
Thank you again! These are the final changes.

**Comment by qliqdev** [comment]
By the way should i change versions in `package.json` from `3.0.1` to `3.0.2`?

**Comment by robingenz** [comment]
> By the way should i change versions in `package.json` from `3.0.1` to `3.0.2`?

No, thats not necessary.

File: ios/Plugin/PrivacyScreenPlugin.swift (hunk 5 above)
Comment (qliqdev) [reply ios/Plugin/PrivacyScreenPlugin.swift]: it was resolved in [ceec1eb](https://github.com/capacitor-community/privacy-screen/pull/51/commits/ceec1ebc2afe3a6aaa82caf8edb7c1975728c46c)

Comment (robingenz) [reply ios/Plugin/PrivacyScreenPlugin.swift]: Sorry, my fault, i mean `onAppDetectCapturing`. I will fix this. Thank you for the great work!

File: ios/Plugin/PrivacyScreenPlugin.swift [hunk 10]
```diff
@@ -50,6 +54,20 @@ public class PrivacyScreenPlugin: CAPPlugin {
         }
     }
 
+    @objc private func onAppDetectCapturing() {
+        guard self.isEnabled else {
+            return
+        }
+
```
Comment (robingenz) [inline ios/Plugin/PrivacyScreenPlugin.swift:61]:
```suggestion
```

**Review by robingenz — CHANGES_REQUESTED** [review]
WIP

**Review by robingenz — APPROVED** [review]

**Comment by robingenz** [comment]
@qliqdev Thank you for your contribution!

**Comment by bayasgalanpf** [comment]
Thanks you, guys for doing this plugin.

i need "screenRecordingStarted" example code of JS version. help me. pls.
i need detect screen recording on IOS.

my below code  not working on iphone device. 

  **PrivacyScreen.addEventListener(
    "screenRecordingStarted",
    function () {
      window.alert("Screen recording");
    },
    false
  );**


Thanks.

**Comment by qliqdev** [comment]
> Thanks you, guys for doing this plugin.
> 
> 
> 
> i need "screenRecordingStarted" example code of JS version. help me. pls.
> 
> i need detect screen recording on IOS.
> 
> 
> 
> my below code  not working on iphone device. 
> 
> 
> 
>   **PrivacyScreen.addEventListener(
> 
>     "screenRecordingStarted",
> 
>     function () {
> 
>       window.alert("Screen recording");
> 
>     },
> 
>     false
> 
>   );**
> 
> 
> 
> 
> 
> Thanks.

Which iOS version you are using?

**Comment by bayasgalanpf** [comment]
iOS version 16.0

**Comment by qliqdev** [comment]
> iOS version 16.0

Looks like everything is ok.

this should be working

in your js/ts file:

```
await PrivacyScreen.addListener('screenRecordingStarted', async () => {
    document.documentElement.classList.toggle(`is-capturing`, true);
});
await PrivacyScreen.addListener('screenRecordingStopped', async () => {
    document.documentElement.classList.toggle(`is-capturing`, false);
});
```

in you global css file: 

```
html.is-capturing{
  -webkit-filter: blur(20px);
  filter: blur(20px);
}
```

**Comment by bayasgalanpf** [comment]
Above code working.  👍
Thank you sir.
